File: StructuralPatterns/Decorator/CSharp/LoggingDecorator.cs
namespace Decorator {
    public class LoggingDecorator : NotifierDecorator {
        public LoggingDecorator(INotifier wrapper) : base(wrapper) {}
        public override void Send(string message) {
            System.Console.WriteLine($"[LOG] Sending notification: {message}");
            base.Send(message);
        }
    }
}

File: StructuralPatterns/Decorator/Go/logging_decorator.go
package main

import "fmt"

type LoggingDecorator struct {
	NotifierDecorator
}

func NewLoggingDecorator(notifier Notifier) *LoggingDecorator {
	return &LoggingDecorator{NotifierDecorator{wrapped: notifier}}
}

func (d *LoggingDecorator) Send(message string) {
	fmt.Printf("[LOG] Sending notification: %s\n", message)
	d.wrapped.Send(message)
}
